Webcrypto encrypt and decrypt tests

engelke opened the pull request:
All algorithms that support the encrypt and decrypt methods are tested.

Commit:
First working test of encrypt (with AES-GCM)

Commit:
Fixed too early return from buffer comparisons.

Commit:
Adding more test data.

Commit:
Cleaned up first test. Now ready to add remaining AES-GCM tests.

Commit:
All GCM tests with 128 bit tags work

Commit:
All GCM successes tested

Commit:
Added failing tests.

Commit:
AES-GCM testing complete

Commit:
Added AES-CBC tests

Commit:
Cleaned up AES-GCM tests

Commit:
Fixed worker imports

Commit:
Worker and timeouts fixed

Commit:
Minor cleanup for vector definitions

Commit:
Added AES-CTR tests

Commit:
Cleaning up and consolidating AES tests

Commit:
Remaining consolidation of AES tests.

Commit:
First RSA test works

Commit:
Testing by roundtrip, and verifying that second encrypt gives different result from first one

Commit:
Plaintext should be too long for bigger hash functions. WIP to fix tests.

Commit:
RSA tests complete.

Commit:
Renamed directory to include decrypt tests, too.

Commit:
RSA-OAEP encrypt and decrypt tested

Commit:
Added failure tests to RSA-OAEP

Commit:
Comprehensive decrypt tests in place.

jimsch commented:
I have not gone thought this in detail yet.  However does the CBC test case have a test for an incorrect padding when decrypting?  One common error is you get the padding of <1, 2, 3, 4, 5> and you remove the last five bytes rather than erroring since all of them are not 5's

engelke commented:
No, I don't test that, but can add it.

Charlie

On Sat, Jun 4, 2016, 1:25 PM Jim Schaad <notifications@github.com> wrote:

> I have not gone thought this in detail yet. However does the CBC test case
> have a test for an incorrect padding when decrypting? One common error is
> you get the padding of <1, 2, 3, 4, 5> and you remove the last five bytes
> rather than erroring since all of them are not 5's
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/3110#issuecomment-223767640>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAE_N7QT5Fz-ArHssdU2Pd8SuTYsvhwiks5qIbUagaJpZM4ItxzV>
> .
>

Commit:
Added CBC padding error tests

engelke commented:
Added CBC padding failure tests.

jimsch commented:
Based on my review I think the following tests are missing:
1  the algorithm of the key and the algorithm of the function do not match
2. algorithms which do not support the encrypt/decrypt operation
3. missing aad for the GCM algorithm
4. Change of the input array after the function call has been made

I am not overly happy with the idea of using a time out to decide if all of the tests have been setup.  It might be better to break the tests up into two steps:
Step 1 - Do the import of all of the keys.
Step 2 - when that is finished (based on Promise.all) then start all of the tests and use a normal done.  Note that this does mean that you need to mark the done as being deferred for the html tests.

engelke commented:
1. No, I didn't test having a CryptoKey that didn't match the algorithm.
That seems like a good idea to me, so I'll add it.

2. This didn't occur to me. I don't think that the recommendation addresses
what should happen in this case. Obviously, it's an error, but what kind of
error?

3. I'll add this.

4. Is this an error? Based on the description, if this happened before the
data was cloned, the operation should succeed, just with the changed data.
If it happened after the data was cloned, it should make no difference. But
which of those two situations occurs seems non-deterministic to me, so
untestable.

Regarding using a timeout for setup - I don't think the tests are doing
that. The only explicit timeout I use is to determine whether the worker
tests (not setup) have finished or not. That's because it seems that I must
call done() explicitly when tests are finished, or else the framework
always marks the tests as TIMEOUT. (The non-worker tests don't need this.)
I suppose I could add every Promise to an array and return that from
run_tests(), and then the worker code could use Promise.all to figure out
when to call done().

I'll think about that, but at first glance, I'm not in favor of it.

I'll make a comment when I think the revised tests are ready for your
review.

Thanks,

Charlie


On Mon, Jun 6, 2016 at 2:06 PM, Jim Schaad <notifications@github.com> wrote:

> Based on my review I think the following tests are missing:
> 1 the algorithm of the key and the algorithm of the function do not match
> 2. algorithms which do not support the encrypt/decrypt operation
> 3. missing aad for the GCM algorithm
> 4. Change of the input array after the function call has been made
>
> I am not overly happy with the idea of using a time out to decide if all
> of the tests have been setup. It might be better to break the tests up into
> two steps:
> Step 1 - Do the import of all of the keys.
> Step 2 - when that is finished (based on Promise.all) then start all of
> the tests and use a normal done. Note that this does mean that you need to
> mark the done as being deferred for the html tests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/3110#issuecomment-224039483>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAE_Nz2h65UMREgyJu_OaOM-tDQY5WYhks5qJGGggaJpZM4ItxzV>
> .
>

jimsch commented:
Regarding issue 4 above, My assumption is that the clone needs to take place before the function call returns.  That is it does not occur at some later time when the promise is being created, instead it is part of the IDL.  This means there is no timing issue involved.  But there does need to be a check that the cloning is occurring and at the correct point in time.

The call to done needs to be made when all of the tests have been started not when they have been completed. This is done implicitly when the load event of an html document is finished (unless it is marked as needing to be done explicitly).  Look at http://testthewebforward.org/docs/testharness-library.html under "determining when all tests are complete"  This means that one can do a pass of all setup work, then create all of the tests after the promises of the first pass run, and finally execute the done function.

I do not believe that you are using the step_timeout function the way it was intended.

Commit:
Added AES-GCM without additional data element tests

Commit:
Properly detect test completion and call done() method.

Commit:
Properly detect test completion for RSA tests, too, and call done() method.

Commit:
Removed step_timeout to detect completion.

Commit:
Checked that data cloning took place for AES

Commit:
Added tests for altered buffers

Commit:
Added test for mismatched algorithm and key

engelke commented:
Jim,

I think I've addressed your issues.

1 the algorithm of the key and the algorithm of the function do not match
- added tests for this for AES. Did not do this for RSA because there is
only one RSA algorithm that does encrypt/decrypt. so there was nothing to
change it to.

2. algorithms which do not support the encrypt/decrypt operation
- have not addressed this. I'm still not clear on the purpose of this.
Wouldn't 1 above handle it, combined with tests that the key has the
encrypt usage? Algorithms that don't support encrypt/decrypt could not get
a key that did support that usage.

3. missing aad for the GCM algorithm
- added

4. Change of the input array after the function call has been made
- added

In addition, instead of using step_timeout for calling done, I keep an
array of all Promises being tested, and use Promise.all to call done when
they've all resolved.

Charlie

jimsch commented:
The check for 1 does not need to check just the variations of the same algorithm.  It would also check that one does not use, for example, an RSA-PSS key in an RSA-OAEP operation. One of the things that people have managed to mess up with JOSE implemenations is that they permit the use of an RSA key object to be used as the key for an HMAC operation.  This check prevents this type of thing from happening assuming you do the key and HMAC operation separately.

I'll review everything again later today.

engelke commented:
Thanks. As for the RSA-PSS/RSA-OAEP issue, there should be no way to create
an RSA-PSS key with the "encrypt" or "decrypt" usage (per tests for
generateKey and importKey), and there are tests here for keys without the
needed usage. So wouldn't that cover things?

Charlie

On Wed, Jun 8, 2016 at 3:36 PM, Jim Schaad <notifications@github.com> wrote:

> The check for 1 does not need to check just the variations of the same
> algorithm. It would also check that one does not use, for example, an
> RSA-PSS key in an RSA-OAEP operation. One of the things that people have
> managed to mess up with JOSE implemenations is that they permit the use of
> an RSA key object to be used as the key for an HMAC operation. This check
> prevents this type of thing from happening assuming you do the key and HMAC
> operation separately.
>
> I'll review everything again later today.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/3110#issuecomment-224703277>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAE_N780V6YpkSYqlhHZhT0bqqgl533Oks5qJxmmgaJpZM4ItxzV>
> .
>

jimsch commented:
This appears to be a case where we cannot see the trees for the forest.  If you look at the procedure, it checks that the name of the algorithm is correct before doing the usage check.  However since the same error is returned for both of them it would not be possible to determine which condition caused it.  The check could be done with an AES key, I agree that it would be odd for user code to try this, but it would check the error condition.  What do you thinK?

engelke commented:
For the RSA issue, either step 8 or 9 would throw the same error, and I
can't think of any test scenario that would let you know which one did it,
so I don't see any reason to try testing this.

For AES, there are tests for this. So are there any outstanding issues you
have for encrypt and decrypt?

Charlie

On Wed, Jun 8, 2016 at 9:32 PM, Jim Schaad <notifications@github.com> wrote:

> This appears to be a case where we cannot see the trees for the forest. If
> you look at the procedure, it checks that the name of the algorithm is
> correct before doing the usage check. However since the same error is
> returned for both of them it would not be possible to determine which
> condition caused it. The check could be done with an AES key, I agree that
> it would be odd for user code to try this, but it would check the error
> condition. What do you thinK?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/3110#issuecomment-224779167>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAE_N4e-jp6CBLntekFbvUzoomrq5fjHks5qJ21JgaJpZM4ItxzV>
> .
>

jimsch commented on WebCryptoAPI/encrypt_decrypt/aes.js:
```diff
@@ -0,0 +1,318 @@
+                return subtle.encrypt(vector.algorithm, vector.key, plaintext)
+                .then(function(result) {
+                    plaintext[0] = 255 - plaintext[0];
+                    assert_true(equalBuffers(result, vector.result), "Should return expected result");
```
This needs to be moved out of the encrypt promise.

engelke replied:
Yeah, I just saw that when I was working on sign/verify. I'll get it fixed
and checked in first thing in the morning. I may be able to get sign and
verify done tomorrow, too.

Charlie

On Thu, Jun 9, 2016 at 6:28 PM, Jim Schaad <notifications@github.com> wrote:

> In WebCryptoAPI/encrypt_decrypt/aes.js
> <https://github.com/w3c/web-platform-tests/pull/3110#discussion_r66533404>
> :
>
> > +            }, "importKey step: " + vector.name);
> > +        });
> > +
> > +        all_promises.push(promise);
> > +    });
> > +
> > +    // Check for successful encryption even if the buffer is changed after calling encrypt.
> > +    passingVectors.forEach(function(vector) {
> > +        var plaintext = copyBuffer(vector.plaintext);
> > +        var promise = importVectorKey(vector, ["encrypt", "decrypt"])
> > +        .then(function(vector) {
> > +            promise_test(function(test) {
> > +                return subtle.encrypt(vector.algorithm, vector.key, plaintext)
> > +                .then(function(result) {
> > +                    plaintext[0] = 255 - plaintext[0];
> > +                    assert_true(equalBuffers(result, vector.result), "Should return expected result");
>
> This needs to be moved out of the encrypt promise.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/3110/files/bdc1e5774a682fb7baec2cc3db973af24b49ac97#r66533404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAE_NyCxNlAUuy25KySKAq6zHRi2CMrsks5qKJOegaJpZM4ItxzV>
> .
>

engelke replied:
Okay, I've fixed this and think we're good to go on these tests.

There is one weirdness that I can't see any way to fix. The tests for
illegal tag lengths fail in Firefox with a JavaScript error on the harness,
but they don't do that in the regular page tests. (They fail in Firefox
either way.)

Charlie

On Thu, Jun 9, 2016 at 6:49 PM, Charles Engelke <charles.engelke@gmail.com>
wrote:

> Yeah, I just saw that when I was working on sign/verify. I'll get it fixed
> and checked in first thing in the morning. I may be able to get sign and
> verify done tomorrow, too.
>
> Charlie
>
> On Thu, Jun 9, 2016 at 6:28 PM, Jim Schaad <notifications@github.com>
> wrote:
>
>> In WebCryptoAPI/encrypt_decrypt/aes.js
>> <https://github.com/w3c/web-platform-tests/pull/3110#discussion_r66533404>
>> :
>>
>> > +            }, "importKey step: " + vector.name);
>> > +        });
>> > +
>> > +        all_promises.push(promise);
>> > +    });
>> > +
>> > +    // Check for successful encryption even if the buffer is changed after calling encrypt.
>> > +    passingVectors.forEach(function(vector) {
>> > +        var plaintext = copyBuffer(vector.plaintext);
>> > +        var promise = importVectorKey(vector, ["encrypt", "decrypt"])
>> > +        .then(function(vector) {
>> > +            promise_test(function(test) {
>> > +                return subtle.encrypt(vector.algorithm, vector.key, plaintext)
>> > +                .then(function(result) {
>> > +                    plaintext[0] = 255 - plaintext[0];
>> > +                    assert_true(equalBuffers(result, vector.result), "Should return expected result");
>>
>> This needs to be moved out of the encrypt promise.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/w3c/web-platform-tests/pull/3110/files/bdc1e5774a682fb7baec2cc3db973af24b49ac97#r66533404>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AAE_NyCxNlAUuy25KySKAq6zHRi2CMrsks5qKJOegaJpZM4ItxzV>
>> .
>>
>
>

Commit:
Fixed time of altering source buffer test

Commit:
Improved test labels for bad tag lengths

jimsch commented:
Ok - The majority of this looks like it is correct and should work.  Charles says it does function so let's merge this and fix any issues that come up later.
